src/metabase/feature_extraction/automagic_dashboards.clj

```diff
@@ -0,0 +1,335 @@
+                           (assoc :aggregation metrics)
+
+                           limit
+                           (assoc :limit limit))}]
+    (when (perms/set-has-full-permissions-for-set?
+           @api/*current-user-permissions-set*
+           (card/query-perms-set query :write))
+      query)))
+
+(defn- card-candidates
+  [context {:keys [metrics filters dimensions score limit order_by] :as card}]
+  (let [filters    (some->> filters
+                            ensure-seq
+                            (map (partial get (:filters context))))
+        dimensions (some->> dimensions
+                            ensure-seq
+                            (map (partial get (:dimensions context))))
+        metrics    (some->> metrics
+                            ensure-seq
+                            (map (partial get (:metrics context))))
+        bindings   (concat filters dimensions metrics)]
+    (when (every? some? bindings)
+      (let [score (* (or score max-score)
+                     (/ (transduce (map :score) + bindings)
+                        max-score
+                        (count bindings)))]
+        (->> (combo/cartesian-product
+              (apply combo/cartesian-product (map :matches filters))
+              (apply combo/cartesian-product (map :matches dimensions))
+              (apply combo/cartesian-product (map :matches metrics)))
+             (mapv (fn [[filters dimensions metrics]]
+                     (when-let [query (build-query (:database context)
+                                                   (-> context :root-table :id)
+                                                   filters
+                                                   metrics
+                                                   dimensions
+                                                   limit
+                                                   order_by)]
+                       (assoc card
+                         :query query
+                         :score score)))))))))
+
+(def ^:private rules-dir "resources/automagic_dashboards")
+
+(defn- load-rules
+  []
+  (->> rules-dir
+       clojure.java.io/file
+       file-seq
+       (filter (memfn ^java.io.File isFile))
+       (map (fn [f]
+              (-> f
+                  slurp
+                  yaml/parse-string
+                  (update :table #(-> %
+                                      (or (let [fname (.getName f)]
+                                            (subs fname 0 (- (count fname) 5))))
+                                      ->type)))))))
+
+(defn- best-matching-rule
+  "Pick the most specific among applicable rules.
+   Most specific is defined as entity type specification the longest ancestor
+   chain."
+  [rules table]
+  (some->> rules
+           (filter #(isa? (:entity_type table) (:table %)))
+           not-empty
+           (apply max-key (comp count ancestors :table))))
+
+(def ^:private ^Integer max-cards 9)
+
+(defn populate-dashboards
```

Comment:
I'd like to have some traces in the logs for the following and probably restructuring a bit to follow the following steps:
0. If adding database wide cards, list the database
1. Output the table name, the type and the guessed filename that applies (to see if generic or which overloaded table type tripped)
2. The creation of each class of artifacts
-- collections
-- dashboards
-- specific dashboard cards
-- (future) metrics